Remora.Commands/Builders/CommandParameterBuilder.cs

```diff
@@ -0,0 +1,453 @@
+        _isGreedy = true;
+
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the parameter to be a named switch.
+    /// </summary>
+    /// <param name="defaultValue">The default value of the switch.</param>
+    /// <param name="shortOrLongName">Either the short name (e.g. '-o') or the long name (e.g. '--switch') of the switch.</param>
+    /// <returns>The builder to chain calls with.</returns>
+    /// <remarks>
+    /// This method will set a default value for the parameter, which will
+    /// implicitly make it optional.
+    /// </remarks>
+    /// <exception cref="InvalidOperationException">Thrown if both tuple values in <paramref name="shortOrLongName"/> are <c>null</c>.</exception>
+    public CommandParameterBuilder IsSwitch(bool defaultValue, OneOf<char, string, (char Short, string Long)> shortOrLongName = default)
+    {
+        _defaultValue = defaultValue;
+        _parameterType = typeof(bool);
+
+        var canBeSwitch = !_attributes.Any(r => r is OptionAttribute);
+
+        if (!canBeSwitch)
+        {
+            throw new InvalidOperationException("A parameter marked as an option cannot be a switch.");
+        }
+
+        var attribute = shortOrLongName.Match
+        (
+            shortName => new SwitchAttribute(shortName),
+            longName => new SwitchAttribute(longName),
+            shortAndLongName => new SwitchAttribute(shortAndLongName.Short, shortAndLongName.Long)
+        );
+
+        _attributes.Add(attribute);
+
+        return this;
+    }
+
+    /// <summary>
+    /// Marks the parameter as being a named option.
+    /// </summary>
+    /// <param name="shortOrLongName">Either the short name (e.g. '-o') or the long name (e.g. '--option') of the option.</param>
+    /// <returns>The builder to chain calls with.</returns>
+    /// <exception cref="InvalidOperationException">Thrown if both values for <paramref name="shortOrLongName"/> are null..</exception>
+    public CommandParameterBuilder IsOption(OneOf<char, string, (char Short, string Long)> shortOrLongName = default)
+    {
+        var canBeOption = !_attributes.Any(r => r is SwitchAttribute);
+
+        if (!canBeOption)
+        {
+            throw new InvalidOperationException("A parameter marked as an option cannot be a switch.");
+        }
+
+        var attribute = shortOrLongName.Match
+        (
+            shortName => new OptionAttribute(shortName),
+            longName => new OptionAttribute(longName),
+            shortAndLongName => new OptionAttribute(shortAndLongName.Short, shortAndLongName.Long)
+        );
+
+        _attributes.Add(attribute);
+
+        return this;
+    }
+
+    /// <summary>
+    /// Adds an attribute to the parameter. Conditions must be added via <see cref="AddCondition"/>.
+    /// </summary>
+    /// <param name="attribute">The attriubte to add.</param>
+    /// <returns>The builder to chain calls with.</returns>
+    public CommandParameterBuilder AddAttribute(Attribute attribute)
```

Comment:
Might be nice to have an overload that takes a generic parameter here with a `new()` constraint - also lets us do compile-time checking for conditions that don't take any parameters (i.e, `where TAttribute : ConditionAttribute` and so on)

Reply:
Curious what kind of checks this would entail